docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,363 @@
+
+sponsors:
+- TBD
+
+reviewers:
+- TBD
+
+approvers:
+- TBD
+
+creation-date: 2022-07-23
+
+---
+
+# Reverse-Proxy Extensions support for Argo CD
+
+Enable UI extensions to use a backend service.
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+* [Goals](#goals)
+* [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Use cases](#use-cases)
+    * [Security Considerations](#security-considerations)
+    * [Risks and Mitigations](#risks-and-mitigations)
+    * [Upgrade / Downgrade](#upgrade--downgrade)
+* [Drawbacks](#drawbacks)
+* [Open Questions](#open-questions)
+
+---
+
+## Summary
+
+Argo CD currently supports the creation of [UI extensions][1] allowing
+developers to define the visual content of the "more" tab inside
+a specific resource. Developers are able to access the resource state to
+build the UI. However, currently it isn't possible to use a backend
+service to provide additional functionality to extensions. This proposal
+defines a new reverse proxy feature in Argo CD, allowing developers to
+create a backend service that can be used in UI extensions. Extensions
+backend code will live outside Argo CD main repository.
+
+## Motivation
+
+The initiative to implement the anomaly detection capability in Argo CD
+highlighted the need to improve the existing UI extensions feature. The
+new capability will required the UI to have access to data that isn't
+available as part of Application's owned resources. It is necessary to
+access an API defined by the extension's development team so the proper
+information can be displayed.
+
+## Goals
+
+The following goals are desired but not necessarily all must be
+implemented in a given Argo CD release:
+
+#### [G-1] Argo CD (API Server) must have low performance impact when running extensions
+
+Argo CD API server is a critical component as it serves all APIs used by
+the CLI as well as the UI. The Argo CD team has no controll over what is
+going to be executed in extension's backend service. Thus it is important
+that the reverse proxy implementation to cause the lowest possible impact
+in the API server while processing high latency requests.
+
+Possible solutions:
+- Implement a rate limit layer to protect Argo CD API server
+- Implement configurable different types of timeouts (idle connection, duration, etc) between Argo CD API server and backend services.
+
+----
+
+#### [G-2] Argo CD admins should be able to define rbacs to define which users can invoke specific extensions
+
+Argo CD Admins must be able to define which extensions are allowed to be
+executed by the logged in user. This should be fine grained by Argo CD
+project like the current rbac implementation.
+
+----
+
+#### [G-3] Argo CD deployment should be independent from backend services
+
+Extension developers should be able to deploy their backend services
+independently from Argo CD. An extension can evolve their internal API and
+deploying a new version shouldn't require Argo CD to be updated or
+restarted.
```

Comment:
I believe that this would be covered in a future enhancement proposal to Argo CD extension where we bind the UI and the backend components together in a cohesive package.